src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs

```diff
@@ -48,5 +52,41 @@ private Matcher CreateMatcher(IReadOnlyList<Endpoint> endpoints)
 
             return builder.Build();
         }
+
+        // Used to tie the lifetime of a DataSourceDependentCache to the service provider
+        public class Lifetime : IDisposable
+        {
+            private readonly object _lock = new object();
+            private DataSourceDependentCache<Matcher> _cache;
+            private bool _disposed;
+
+            public DataSourceDependentCache<Matcher> Cache
+            {
+                get => _cache;
+                set
+                {
+                    lock (_lock)
+                    {
+                        if (_disposed)
+                        {
+                            value?.Dispose();
```

Comment:
This will allow operating on a disposed instance. Should it rather throw an ObjectDisposedException in this case?